### PR TITLE
Password-protected Tokens

### PR DESCRIPTION
**Pull request by Magnus-Kuhn**
# Readiness checklist

-   [X] I added/updated tests.
-   [X] I ensured that the PR title is good enough for the changelog.
-   [X] I labeled the PR.

### CODE REVIEW

File: packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenQRCodeForOwnRelationshipTemplate.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenQRCodeForOwnRelationshipTemplate.ts:56]: I'd prefer an isEqual method on the PasswordProtection object. Comparing classes with lodash's isEqual should be avoided.

Comment (Magnus-Kuhn) [reply packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenQRCodeForOwnRelationshipTemplate.ts:56]: I see, then some isEqual checks in the AttributeController need a JSON conversion first, I'm on it

**Comment by Magnus-Kuhn** [comment]
Some TODOs not in this PR are:
* Allowing querying for password-protection in the runtime (I forgot it with the templates, so I thought to do them separately for both templates and tokens)
* Completing the refactoring where validations for `expiresAt` and `passwordProtection` are moved into the SchemaValidator by renaming the SchemaValidator

File: packages/runtime/src/useCases/common/validation/SchemaValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/SchemaValidator.ts:31]: What's going on here? Why do you put logic for some usescases in the baseclass?! It would be way cleaner if you just extend the SchemaValidator and put it there any only use the new extended version where required.

File: packages/runtime/src/useCases/transport/account/LoadItemFromTruncatedReference.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/account/LoadItemFromTruncatedReference.ts:1]: you forgot to pass the password down to the `loadPeerRelationshipTemplateByTokenContent` and `getOrLoadFile` methods. I'm **very** curious why this didn't popped up in a test, please double check that you test this usecase with a double-password-protected item.

File: packages/runtime/src/useCases/transport/files/UploadOwnFile.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/files/UploadOwnFile.ts:64]: I don't like, that you squeezed in a major refactoring into this PR. This has nothing to do with the actual PR.

File: packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts:66]: why did you rename this? This results in a way less clean usage below.

File: packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenQRCodeForOwnRelationshipTemplate.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenQRCodeForOwnRelationshipTemplate.ts:65]: see above

File: packages/runtime/src/useCases/transport/relationshipTemplates/LoadPeerRelationshipTemplate.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/relationshipTemplates/LoadPeerRelationshipTemplate.ts:58]: please write a test for this, it's a pity that you didn't discovered this in the other PR by writing proper tests.

File: packages/runtime/test/lib/testUtils.ts
Comment (jkoenig134) [inline packages/runtime/test/lib/testUtils.ts:163]: when just calling this password, you completely loose the possibility to properly test this for example with PINs. Why not allow to pass an object with password and passwordIsPin? Would also make the usage below cleaner.

File: packages/runtime/test/transport/account.test.ts
Comment (jkoenig134) [inline packages/runtime/test/transport/account.test.ts:95]: Tests should be independent, please create a reference for each test and don't clutter up the test more than it is currently.

File: packages/runtime/test/transport/relationshipTemplates.test.ts
Comment (jkoenig134) [inline packages/runtime/test/transport/relationshipTemplates.test.ts:1]: I said it many times before - when adding a new feature it's an antipattern to just stuff in that new feature in old tests and weaken them by this.

We should test all combinations, what you do doesn't help with that.

Comment (jkoenig134) [reply packages/runtime/test/transport/relationshipTemplates.test.ts:1]: what you do:

1. test a password-protected template in a password-protected token

what you should do:

1. test a non-protected template in a password-protected token
2. test a password-protected template in a password-protected token
3. test a password-protected template in a non-protected token (is that even possible? how should we handle this?)

File: packages/runtime/test/transport/tokens.test.ts
Comment (jkoenig134) [inline packages/runtime/test/transport/tokens.test.ts:195]: please create a file for each test. Then you are way more flexible to test all combinations (see above for more context)

File: packages/transport/src/core/types/PasswordProtection.ts
Comment (jkoenig134) [inline packages/transport/src/core/types/PasswordProtection.ts:36]: `compare` is not a good wording. wtb just `creationParameters` to match the function naming?

File: packages/transport/src/core/types/PasswordProtection.ts
Comment (jkoenig134) [inline packages/transport/src/core/types/PasswordProtection.ts:40]: see above: `newPasswordProtection` would be a better naming for `compare`. Also the method name could benefit from the word `Protection` at the end to make clearer that this is not ONLY about passwords and also about pins.

File: packages/transport/src/core/types/PasswordProtection.ts
Comment (jkoenig134) [inline packages/transport/src/core/types/PasswordProtection.ts:44]: as this function calls the other private function the private function has to be BELOW this. You can read more about this in our coding guidelines.

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

File: packages/runtime/src/useCases/transport/account/LoadItemFromTruncatedReference.ts
Comment (Magnus-Kuhn) [reply packages/runtime/src/useCases/transport/account/LoadItemFromTruncatedReference.ts:1]: Password protection for files only works via tokens, so `getOrLoadFile` doesn't get a password - and the test did not have the double password-protection, so the error in the `relationshipTemplateTokenContent` was uncaught

Comment (jkoenig134) [reply packages/runtime/src/useCases/transport/account/LoadItemFromTruncatedReference.ts:1]: I see.

File: packages/runtime/src/useCases/common/validation/SchemaValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/SchemaValidator.ts]:
```suggestion
export class SchemaValidator<T> implements IValidator<T> {
```

File: packages/runtime/src/useCases/common/validation/GenericInputValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/GenericInputValidator.ts]: I'm very confused, could you elaborate the naming here? Why is this InputValidator more GENERIC than the class that it extends?

Comment (jkoenig134) [reply packages/runtime/src/useCases/common/validation/GenericInputValidator.ts]: Also, Schema is missing in the name.

File: packages/runtime/src/useCases/transport/files/CreateTokenForFile.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/files/CreateTokenForFile.ts]: so this is perfectly fine - but does not belong in this PR.

Comment (jkoenig134) [reply packages/runtime/src/useCases/transport/files/CreateTokenForFile.ts]: as it's done already keep it.

Comment (jkoenig134) [reply packages/runtime/src/useCases/transport/files/CreateTokenForFile.ts]: Hm. Now I see another issue with the implementation - which is **exactly** the reason why this should be in another PR

1. use CoreDate instead of DateTime
2. use `CoreDate#isExpired()` to check against `DateTime.now()`

File: packages/runtime/src/useCases/transport/files/CreateTokenQRCodeForFile.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/files/CreateTokenQRCodeForFile.ts]: see other comment

File: packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts:64]: Hm, a more broader question that get's in my head while reading this:

Theoretically it's possible now to put a passwordProtected template inside of a non-protected token

1. do we want that?
2. will we run into issues with that? I don't really know how to handle this..

File: packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts:8]:
```suggestion
export class TokenAndTemplateCreationValidator<
    T extends {
        expiresAt?: ISO8601DateTimeString;
        passwordProtection?: {
            password: string;
            passwordIsPin?: true;
        };
    }
> extends SchemaValidator<T> {
```

File: packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts:20]:
```suggestion
        if (input.expiresAt && CoreDate.from(input.expiresAt).isExpired()) {
            validationResult.addFailure(new ValidationFailure(RuntimeErrors.general.invalidPropertyValue(`'expiresAt' must be in the future`), "expiresAt"));
        }
```

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

**Comment by jkoenig134** [comment]
It seems like you have committet some things during my review - please make sure to read all comments, also if they are marked as outdated as this can be a false positive.

**Comment by jkoenig134** [comment]
Also this PR conflicts.

File: packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts
Comment (Magnus-Kuhn) [reply packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts:64]: `matchesInputForNesPasswordProtection` will return false if the new password protection is undefined - and for a protected template inside an unprotected token, the reference from the token would not reveal that a password is required. Hence I want to block that

Comment (jkoenig134) [reply packages/runtime/src/useCases/transport/relationshipTemplates/CreateTokenForOwnRelationshipTemplate.ts:64]: Oh, I misunderstood this. Looks good then.

File: packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts
Comment (jkoenig134) [inline packages/runtime/src/useCases/common/validation/TokenAndTemplateCreationValidator.ts:34]: if I look up at the other error I think you should use `RuntimeErrors.general.invalidPropertyValue` instead of a new error

File: packages/runtime/test/transport/tokens.test.ts
Comment (jkoenig134) [reply packages/runtime/test/transport/tokens.test.ts]: I think you misunderstood this comment. I really meant just an **enmeshed file** not test.ts files, but I think it's ok like it is now?
@Milena-Czierlinski wdym?

File: packages/runtime/src/useCases/common/validation/GenericInputValidator.ts
Comment (jkoenig134) [reply packages/runtime/src/useCases/common/validation/GenericInputValidator.ts]: This is not fully resolved.

**Review by jkoenig134 — APPROVED** [review]
LGTM. Please wait for another review by @britsta or @Milena-Czierlinski

**Review by Milena-Czierlinski — APPROVED** [review]
Please make sure to also test for the error message of `recordNotFound` once the Backbone has adjusted it.

**Comment by jkoenig134** [comment]
@Milena-Czierlinski typically we don't test error messages, only codes.

**Review by jkoenig134 — APPROVED** [review]

**Comment by jkoenig134** [comment]
@Magnus-Kuhn anything that blocks this from merging?

**Pull request merged**
